src/Helpers/TenantWrapper.php

```diff
@@ -50,6 +50,56 @@ public static function with(Tenant $tenant): self
      */
     public function getSpEntityId(): string
     {
-        return $this->tenant->sp_entity_id_override ?: URL::route('saml.metadata', ['uuid' => $this->tenant->uuid]);
+        return $this->getUrlWithDomainOverrideIfConfigured('saml.metadata');
+    }
+
+    /**
+     * Use this to determine the ACS URL for a tenant.
+     *
+     * 1. Returns default ACS URL *UNLESS* id_app_url_override is filled in (non-empty string),
+     * 2. ... in which case it will return the ACS path appended to that value instead of
+     *    what URL::route uses (APP_URL at command line, request HOST during a request)
+     *
+     * Q: Why not just make this Tenant::getAcsUrlAttribute?
+     * A: Because Eloquent models hate unit testing
+     *
+     * @return string ACS URL (full path)
+     */
+    public function getAcsUrl(): string
+    {
+        return $this->getUrlWithDomainOverrideIfConfigured('saml.acs');
+    }
+
+    /**
+     * Use this to determine the Single Logout Service URL for a tenant.
+     *
+     * 1. Returns default SLS URL *UNLESS* id_app_url_override is filled in (non-empty string),
+     * 2. ... in which case it will return the SLS path appended to that value instead of
+     *    what URL::route uses (APP_URL at command line, request HOST during a request)
+     *
+     * @return string SLS URL (full path)
+     */
+    public function getSlsUrl(): string
+    {
+        return $this->getUrlWithDomainOverrideIfConfigured('saml.sls');
+    }
+
+    /**
+     * By default (id_app_url_override not configured), returns default route URL
+     * (which, for URL::route, is a /-relative path, not an absolute including domain)
+     *
+     * When id_app_url_override is configured, the route *path* appended to that id_app_url_override value
+     *
+     * @param string $routeName Valid route — intended to be one of the saml.* routes
+     * @return string path or full URL of that route
+     */
+    private function getUrlWithDomainOverrideIfConfigured(string $routeName): string
+    {
+        if ($this->tenant->id_app_url_override) {
+            $absolute = false;
+            return $this->tenant->id_app_url_override . URL::route($routeName, ['uuid' => $this->tenant->uuid], $absolute);
+        }
+
+        return URL::route($routeName, ['uuid' => $this->tenant->uuid]);
```

Comment:
For readability can we make the params their own variable and remove the duplication. I also think it might be worth considering adding some tolerance regarding slashes on override value.

```suggestion
        $params = [ 'uuid' => $this->tenant->uuid ];
        $path = URL::route($routeName, $params);
        
        // Relative path without URL override
        if ($this->tenant->id_app_url_override === null) {
            return $path;
        }

        // Absolute urls with override
        $absolute = false;
        $url = rtrim($this->tenant->id_app_url_override, '/')

        return $url . URL::route($routeName, $params, $absolute);
```

Reply:
> For readability can we make the params their own variable and remove the duplication.

I had refactored to that because sonarcloud complained 😆

Reply:
agreed, using $params, and rtrim /
Added a test case to repro the slash problem before adding that line.